designs/package_during_deploy.md

```diff
@@ -0,0 +1,175 @@
+
+This is driven by the following workflow.
+
+`sam build` -> `sam package --s3-bucket ... --s3-prefix ... --output-template-file packaged.yaml`.
+
+This workflow builds the requisite code and packages those built artifacts into an s3 bucket, optionally under a given s3 prefix.
+
+If a developer can optionally cut through this process without requiring an explicit `package` command, but rather have `sam deploy` package to a given s3 bucket, it cuts the number of steps before needing to deploy and test in the cloud.
+
+This also reduces friction in the `author` and `test` loop.
+
+Ideal end state.
+
+`sam build` -> `sam deploy ..... --s3-bucket ..`
+
+
+What will be changed?
+---------------------
+
+Addition of extra parameters that are currently supported by `sam package` over to `sam deploy` with all of them being optional.
+
+Additional parameters that need to be on `sam deploy` that are not on `sam package`.
+
+* `--metadata`
+* `--use-json`
+
+Parameters that dont need to be added.
+
+* `--output-template-file`
+ * 	An explicit `output-template-file` is created on the fly during packaging in the deploy phase.
+
+If the expectation is to package and deploy in one command, One can now do.
+
+`sam deploy --stack-name sam-package-on-deploy --capabilities CAPABILITY_IAM --s3-bucket sam-package-bucket`
+
+There is no explicit need to pass in a `--template-file` parameter, if one is not passed in it to defaults to trying to find the template.yaml that was generated by `sam build` which is located at `.aws-sam/build/template.yaml`
+
+The old method of deploying pre-packaged artifacts will continue to work as before.
+
+* `sam deploy --template-file packaged.yaml --stack-name sam-package-on-deploy --capabilities  CAPABILITY_IAM`
+
+Future of `sam package`?
+---------------------
+
+* `sam package` will continue to exist in the state it is today and will continue to be improved upon separately.
+
+Success criteria for the change
+-------------------------------
+
+User do not require to run `sam package` as part of their author test loop, except for CI/CD purposes, where `sam package` can be run and the packaged template file can be passed to cloudformation deploy actions.
+
+
+Out-of-Scope
+------------
+
+The s3 bucket where the packaged artifacts go is not abstracted in this design. In the future, the s3 bucket could be specified via a configuration file.
+
+This is currently in design in : https://github.com/awslabs/aws-sam-cli/pull/1503
+
+User Experience Walkthrough
+---------------------------
+
+`sam build` -> `sam deploy`
+
+`sam build` -> `sam package` -> `sam deploy`
+
+Provide identical experiences in terms of a deploying the same stack, with exactly same artifacts.
+
+
+Implementation
+==============
+
+CLI Changes
+-----------
+
+* Add new arguments `--metadata`, `--use-json` and modify existing `--template-file` to look for a default `template.yaml`  that exists under `.aws-sam/build/`
```

Comment:
What value does --use-json provide in the deploy experience?

Reply:
If a given template file is above 51,200 bytes. that corresponding template also gets uploaded to an s3 bucket, if `--use-json` is provided, that uploaded template_file will be in the json format.